Fix SSL_shutdown error handling

marshall-lee opened the pull request:
`SSL_shutdown()` call may result in error and in OpenSSL every error should be consumed by the valler via `SSL_get_error()` or a similar func.

`MiniSSL` implementation was breaking this convention so it resulted in cryptic bugs like SSL clients (not related to Puma in any way) return errors with text like `"SSL_read: shutdown while in init"` that is initially generated by the Puma reactor but not consumed by it.

Such bugs are really hard to debug...

### Description
Thank you for contributing! You're the best.

We can read your code, so consider leaving some comments here that are more about your motivations and decision making. Some things that may be helpful to address in your description:

- What original problem led to this PR?
- Are there related issues / prior discussions?
- What alternatives have been tried? Does this supersede previous attempts?
- Why do you make the choices you did? What are the tradeoffs?

### Your checklist for this pull request
<!--- Go over all the following points, and put an `x` in all the boxes that apply. You can delete or just add an X if you think its not applicable. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have reviewed the [guidelines for contributing](../blob/master/CONTRIBUTING.md) to this repository.
- [x] I have added (or updated) appropriate tests if this PR fixes a bug or adds a feature.
- [ ] My pull request is 100 lines added/removed or less so that it can be easily reviewed.
- [ ] If this PR doesn't need tests (docs change), I added `[ci skip]` to the title of the PR.
- [ ] If this closes any issues, I have added "Closes `#issue`" to the PR description or my commit messages.
- [ ] I have updated the documentation accordingly.
- [x] All new and existing tests passed, including Rubocop.

schneems commented:
Thanks for the PR! The C extension is probably the hardest area to modify, if you can give us some help validating your change I think this looks like a good quality of life improvement:

Can you give me an example of a change in error output before and after? Ideally, if you have a reproduction that lets me exercise and view the error on my own that would be even better https://www.codetriage.com/reproduction.

I tried pulling in the added tests into master/main I'm unable to use them to reproduce an error as they rely on additions from this PR:

```
  1) Error:
TestPumaServerSSLClientCloseError#test_client_close_raises_ssl_error_in_http10:
NoMethodError: undefined method 'errors' for an instance of SSLLogWriterHelper
    test/test_puma_server_ssl.rb:77:in 'TestPumaServerSSLClientCloseError#assert_ssl_client_error_match'
    test/test_puma_server_ssl.rb:83:in 'TestPumaServerSSLClientCloseError#test_client_close_raises_ssl_error_in_http10'

  2) Error:
TestPumaServerSSLClientCloseError#test_client_both_read_and_close_raise_ssl_error:
NoMethodError: undefined method 'errors' for an instance of SSLLogWriterHelper
    test/test_puma_server_ssl.rb:77:in 'TestPumaServerSSLClientCloseError#assert_ssl_client_error_match'
    test/test_puma_server_ssl.rb:92:in 'TestPumaServerSSLClientCloseError#test_client_both_read_and_close_raise_ssl_error'

26 runs, 48 assertions, 0 failures, 2 errors, 1 skips
```

Does this change modify or break an API?

marshall-lee commented:
@schneems 
> I tried pulling in the added tests into master/main I'm unable to use them to reproduce an error as they rely on additions from this PR:
> Does this change modify or break an API?

No just cherry pick the change in `test/helper.rb` as well

marshall-lee commented:
@schneems 

> Can you give me an example of a change in error output before and after? Ideally, if you have a reproduction that lets me exercise and view the error on my own that would be even better https://www.codetriage.com/reproduction.

Well it's tricky.

Before this change, Puma was not printing anything in the logs if `SSL_shutdown()` resulted in error because the error is handled.

However, the only bug I was able to catch is when `SSL_shutdown()` being call too early when the SSL connection is not fully established. This is so called `"shutdown while in init"` error but in this PR it's also ruled out by `SSL_in_init()` check. So with this PR the result is the same: `"shutdown while in init"` will never be printed in the logs.
But I think it's legitimate. Printing an error like `"shutdown while in init"` is too verbose, lets not do this anyway, it's would be a real noise in the logs. Better to silence it with a legitimate `SSL_in_init()` check.

To reproduce at least something you could apply the following patch on current `master` and run some tests:
```
diff --git a/ext/puma_http11/mini_ssl.c b/ext/puma_http11/mini_ssl.c
index 2af28073..5e104541 100644
--- a/ext/puma_http11/mini_ssl.c
+++ b/ext/puma_http11/mini_ssl.c
@@ -663,11 +663,21 @@ VALUE engine_shutdown(VALUE self) {
   ERR_clear_error();
 
   ok = SSL_shutdown(conn->ssl);
-  if (ok == 0) {
+  // See
+  // https://github.com/openssl/openssl/blob/openssl-3.5.2/ssl/ssl_lib.c#L2792-L2797
+  // for description of SSL_shutdown return values.
+  switch (ok) {
+  case 0:
+    // "close notify" alert is sent by us.
     return Qfalse;
+  case 1:
+    // "close notify" alert was received from peer.
+    return Qtrue;
+  default:
+    raise_error(conn->ssl, ok);
   }
 
-  return Qtrue;
+  return Qnil;
 }
 
 VALUE engine_init(VALUE self) {
```

This is nearly the same as changes from my PR except the `SSL_in_init` check. I expect you to see a bunch of `shutdown while in init` failures in the tests.

marshall-lee commented:
Besides of `"shutdown while in init"` (which this PR makes silent by `SSL_in_init()` check) I don't know any other way how to make `SSL_shutdown()` return an error. So i'm sorry, it's not easy to break it down into before / after behavior.

marshall-lee commented:
Well one could ask why do all this if nothing was changed anyway in terms of what error are reported in the logs.

The problem is some OpenSSL resulted in error, it's a caller responsibility to **consume this error** from the **error queue**. 
See the relevant piece of code from Ruby `openssl` gem: https://github.com/ruby/openssl/blob/v3.3.0/ext/openssl/ossl_ssl.c#L2170-L2176
Puma's `raise_error` function does this too: https://github.com/puma/puma/blob/v6.6.1/ext/puma_http11/mini_ssl.c#L585

marshall-lee commented:
@schneems 

Two tests I added in `TestPumaServerSSLClientCloseError` don't really reproduce the root issue I faced.

I added them just because I also introduced `close_client_safely` method that handles `client.close` exceptions. I had to do this because I inserted `raise_error` into the `client.close` internals.

MSP-Greg commented:
> I had to do this because I inserted `raise_error` into the `client.close` internals.

Yes, I didn't notice that right away.  This PR seems correct.  Reviewing OpenSSL man pages, may add a few tests to show that nothing is logged.

marshall-lee commented:
> Also, rather than extract the code to close_client_safely, maybe the following (with 2 added ssl related lines)?

It was extracted cause it has more than one call sites

MSP-Greg commented:
JFYI, as to missing that, the new HTTP_10 test is called from `Server#process_client`, and the HTTP_11 test is called from `Server#reactor_wakeup`.  The two call locations...

marshall-lee commented:
@MSP-Greg 

> Can it be added to Client.close? It's an internal object, and its already got the io wrapped in a rescue block?

`Client` does not have server's `@log_writer`.

MSP-Greg commented:
@schneems

Not sure what you checked.  Some of this code is from 25-July-2016, OpenSSL 1.0.2h was the current version.  I reviewed this with a few current OpenSSL man pages (see https://docs.openssl.org/master/man3/SSL_shutdown/).

LGTM.

marshall-lee commented:
@schneems @MSP-Greg Hi! Is there anything else I can do for this to be merged?
